[OpenMP] Fix for two instances of multiple privatization

kiranchandramohan opened the pull request:
Privatization was happening more than once in two different situations.
1. If private clauses are part of a combined construct then
privatization only needs to happen once. In the code base currently this
happens only for parallel_do. Issue is fixed by not performing
privatization while creating the parallel op.

2. Loop index variables are privatized by default if they occur in a
parallel region. They were further privatized if part of a private
clause. Fixed the issue by not performing privatization if it is part
of a private clause.

Note: Also switching to American spelling for privatize.

schweitzpgi commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -202,7 +206,7 @@ genOMP(Fortran::lower::AbstractConverter &converter,
 static void createParallelOp(Fortran::lower::AbstractConverter &converter,
                              Fortran::lower::pft::Evaluation &eval,
-                             const Directive &directive) {
+                             const Directive &directive, bool isCombined) {
```
I think `isCombined` could be converted to a template argument here. (Quick scan suggests it is always passed as a constant.)

kiranchandramohan replied:
OK, that sounds like a good idea. Done.

schweitzpgi commented on flang/test/Lower/OpenMP/omp-parallel-private-clause.f90:
```diff
@@ -41,3 +41,22 @@ subroutine private_clause(arg1, arg2, arg3, arg4)
+!FIRDialect-DAG: omp.parallel
+!FIRDialect-DAG: %{{.*}} = fir.alloca i32 {{{.*}}uniq_name = "{{.*}}j"}
+!FIRDialect-DAG: %{{.*}} = fir.alloca i32 {{{.*}}uniq_name = "{{.*}}x"}
+!FIRDialect-DAG: omp.wsloop
```
Would using `mlir/utils/generate-test/checks.py` work here for tighter checking?

kiranchandramohan replied:
OK. I have now use the script to generate checks.

schweitzpgi reviewed (approved):
+1

kiranchandramohan commented:
@schweitzpgi can this PR be merged?

clementval commented:
I get failure with this PR. 

```
FAIL: Flang :: Lower/OpenMP/omp-parallel-private-clause-fixes.f90 (1261 of 1263)
******************** TEST 'Flang :: Lower/OpenMP/omp-parallel-private-clause-fixes.f90' FAILED ********************
Script:
--
: 'RUN: at line 3';   bbc -fopenmp -emit-fir /local/home/vclement/llvm-project/flang/test/Lower/OpenMP/omp-parallel-private-clause-fixes.f90 -o - |    /local/home/vclement/llvm-project/build/bin/FileCheck /local/home/vclement/llvm-project/flang/test/Lower/OpenMP/omp-parallel-private-clause-fixes.f90
--
Exit Code: 1

Command Output (stderr):
--
/local/home/vclement/llvm-project/flang/test/Lower/OpenMP/omp-parallel-private-clause-fixes.f90:11:10: error: CHECK: expected string not found in input
! CHECK: %[[VAL_0:.*]] = fir.alloca i32 {uniq_name = "j"}
         ^
<stdin>:5:16: note: scanning from here
 omp.parallel {
               ^
<stdin>:6:2: note: possible intended match here
 %3 = fir.alloca i32 {bindc_name = "j", pinned}
 ^

Input file: <stdin>
Check file: /local/home/vclement/llvm-project/flang/test/Lower/OpenMP/omp-parallel-private-clause-fixes.f90

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            1: func @_QPmultiple_private_fix(%arg0: !fir.ref<i32>) { 
            2:  %0 = fir.alloca i32 {bindc_name = "i", uniq_name = "_QFmultiple_private_fixEi"} 
            3:  %1 = fir.alloca i32 {bindc_name = "j", uniq_name = "_QFmultiple_private_fixEj"} 
            4:  %2 = fir.alloca i32 {bindc_name = "x", uniq_name = "_QFmultiple_private_fixEx"} 
            5:  omp.parallel { 
check:11'0                    X error: no match found
            6:  %3 = fir.alloca i32 {bindc_name = "j", pinned} 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:11'1      ?                                               possible intended match
            7:  %4 = fir.alloca i32 {bindc_name = "x", pinned, uniq_name = "_QFmultiple_private_fixEx"} 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            8:  %c1_i32 = constant 1 : i32 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            9:  %5 = fir.load %arg0 : !fir.ref<i32> 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           10:  %c1_i32_0 = constant 1 : i32 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           11:  omp.wsloop (%arg1) : i32 = (%c1_i32) to (%5) step (%c1_i32_0) inclusive { 
check:11'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            .
            .
            .
>>>>>>

--

********************
********************
Failed Tests (1):
  Flang :: Lower/OpenMP/omp-parallel-private-clause-fixes.f90
```

kiranchandramohan commented:
Thanks @clementval. Required some changes after the pinned attribute patch (https://github.com/flang-compiler/f18-llvm-project/pull/1065). I have now updated the test and squashed the commits.

clementval commented:
@kiranchandramohan Thanks. Let me build it on my side and if ok I'll merge it.
